### PR TITLE
Enable a way to checkout Pull Requests from remote refs

### PR DESCRIPTION
**Pull request by mhartkorn**
This Pull Request adds a new feature described in #1655. Having this feature would be very nice, especially for Go projects where the imports are often global and need to be manually changed to test Pull Requests. It is work in progress and needs feedback.

To make it work, the patch from https://github.com/gogits/git-module/pull/5 needs to be applied to the imports. This is the reason why the Travis-CI build is currently failing.

##### How it works

When a user pushes commits to a branch gogs will find out if it's part of an open Pull Request. If it is, gogs will create a temporary remote to the base repository and push the changes to `refs/pull/<pr-number>/head`. Everyone with permission can then checkout the code from the Pull Request as described here: [Checking out pull requests locally](https://help.github.com/articles/checking-out-pull-requests-locally/)

##### Current status
- [x] Push commits to the base repository into `refs/pull/<pr-number>/head`
- [x] Block cyclic dependencies (not really a problem because of git objects)
- [x] ~~Forbid manual pushes to `refs/pull/*`~~ (handle it like Gitlab and simply override `refs/pull/<id>/head` if already exists)

Please give feedback on the method I'm using as well as on the coding style.

### CODE REVIEW

**Comment by bkcsoft** [comment]
Don't like the last checkbox, since that could still be useful for cherrypicking fixes.

Example:
We have `master`, which has merged pr#2 after creating the tag `v1.0` and no new `v1.1` has been made after the merge. Now if my prod-env needs pr#2 I can still pull (and merge) that one, without having to pull and merge `master`

**Comment by bkcsoft** [comment]
Otherwise :100:

**Comment by Unknwon** [comment]
Quick question: why do we need post-receive instead of update hook?

**Comment by mhartkorn** [comment]
The `update` hook is called before the commits are added to the repository, so if I start a push there, the most recent commits are missing. I searched for the most fitting hook for this feature and `post-receive` was named as the most fitting to do deployments/send the code somewhere.

Maybe the update hook can tweaked for this and I missed something. I would rather have it in the `update` hook but I wasn't able to find out how to do it.

**Comment by mhartkorn** [comment]
@bkcsoft: Good idea, I will remove the last point.

File: cmd/pull.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// Copyright 2015 The Gogs Authors. All rights reserved.
```
Comment (andreynering) [inline cmd/pull.go]: Should be 2016?

**Comment by andreynering** [comment]
:+1: on this. It's very useful.

File: cmd/pull.go (hunk 1 above)
Comment (mhartkorn) [inline cmd/pull.go]: Whoops, I will change it in the next commit.

**Comment by Unknwon** [comment]
@mhartkorn 

>  The update hook is called before the commits are added to the repository, so if I start a push there, the most recent commits are missing.

Update hook itself receives latest commits already... does not need to check with repository again.

**Comment by Unknwon** [comment]
Not commits, but the start and end commits SHA, see https://github.com/gogits/gogs/blob/master/cmd/update.go#L47-L49

> How it works

When a user pushes commits to a branch gogs will find out if it's part of an open Pull Request. If it is, gogs will create a temporary remote to the base repository and push the changes to refs/pull/<pr-number>/head. Everyone with permission can then checkout the code from the Pull Request as described here: Checking out pull requests locally

Not sure, just an idea: why just base repository simply add a remote points to the branch? Never tried, I'll test out.

**Comment by mhartkorn** [comment]
@Unknwon I'm not 100% sure but as far as I understood the git docs, git is still able to reject the push after `update`, isn't it?

**Comment by Unknwon** [comment]
>  I'm not 100% sure but as far as I understood the git docs, git is still able to reject the push after update, isn't it?

Yes, why does it matter

**Comment by mhartkorn** [comment]
@Unknwon I think Pull Requests should only have commits in the base repo (as `refs/pull/*`) that are also in the head repo.

**Comment by Unknwon** [comment]
> @Unknwon I think Pull Requests should only have commits in the base repo (as refs/pull/*) that are also in the head repo.

Why it needs to be pushed?

**Comment by Unknwon** [comment]
I think we may need to come up a feature design to share our understandings?

**Comment by mhartkorn** [comment]
I can give you a rundown on the thought process I had while I programmed this feature.

##### How it currently works in the code

- Receive commit from client thorugh HTTP(S)/SSH
- Git applies the commit to the repository
- Gogs gets notified through the `post-receive` hook that new commits are in the repository
- Gogs checks if the branch of the commits currently is associated with in an open Pull Request
- If yes:
  - Locate the base repository via the file system (using `pwd` and relative paths)
  - Create a temporary remote in the head repository to the base repository
  - Push the branch that is associated with the Pull Request to the base repository into the ref `refs/pull/<pull-request-number>/head`
  - Remove the temporary remote
- If no:
  - Do nothing

##### Why I used the `post-receive` instead of `update`

First I thought I can use in `update`, which is called for each pushed ref with the parameters `old-rev`, `new-rev`, and `ref-name`. But as far as I know `update` does not guarantee that the pushed refs are actually able to be applied to the repository. So if somebody pushes to a wrong branch by mistake, the push gets eventually rejected by git. So I thought, that `post-receive` would be a better place for this. I also noticed that it was not so easy for me to actually get the contents of the commits in `update` to send them to the base repository.

I would be glad if I am wrong here, and can actually use `update` because it reduces complexity.

**Comment by Unknwon** [comment]
Thanks for the great detail!

> So if somebody pushes to a wrong branch by mistake, the push gets eventually rejected by git.

I don't think `update` hook would be called in this case.

**Comment by Unknwon** [comment]
> Push the branch that is associated with the Pull Request to the base repository into the ref refs/pull/<pull-request-number>/head

Why this step is required?

**Comment by mhartkorn** [comment]
> In my understanding, we could just update the refs/pull/ID/head record. When user tries to pull, Gogs knows which repository and branch this record points to and just response data from there.

That was my first thought, too. But I wasn't able to do it because git was always complaining that it was not able to find the commit IDs. Maybe it would be possible if all git related thing were tunneled through a service that then identifies specific use cases (checkout of pull requests, cyclic dependencies between PRs) and then handles them accordingly. This would increase the complexity a lot, though.

I researched how Gitlab is handling their *merge requests* and as far as I can see, they do the same as my current implementation does. They just add a so called *gitlab-shell* that every git related data is getting through to handle some stuff (like denying a manual push to `refs/merge-requests/<id>/head`).

Regarding the problem with cyclic dependencies I came to the conclusion that they are not really a problem. Git stores everything as objects and if there are cyclic dependencies they are just references between each other that do not significantly increase the size of the repository.

**Comment by mhartkorn** [comment]
I did a bit more research and as far as I can see, my implementation in the current state is the minimal version that does not need a permanently running service that handles git access. If there was such a service between the client and git, there would be more possibilities to restrict access to e.g. the refs for Pull Requests. An example for such a service is the [gitlab-shell](https://gitlab.com/gitlab-org/gitlab-shell) that can transparently handle permissions.

**Comment by Unknwon** [comment]
OK... I think I can agree on push content, but I'm trying to avoid using `post-receive` hook.

What I would do is adding push actions [here](https://github.com/gogits/gogs/blob/master/routers%2Frepo%2Fpull.go#L646).

This trigger won't be called if the user push action was not successful (decline to push, etc).

**Comment by mhartkorn** [comment]
@Unknwon Ok, I will look into it and add a commit to this PR in the next few days.

**Comment by mhartkorn** [comment]
@Unknwon Using `TriggerTask` was a very good idea. The code is much cleaner and more robust now.

**Comment by Unknwon** [comment]
Great, I'll review the push step detials ASAP.

**Comment by JohnMaguire** [comment]
May I ask why you are trying to avoid using `post-receive`? I think it makes more sense here:

`update`:
> This hook is invoked by git-receive-pack on the remote repository, which happens when a git push is done on a local repository. Just before updating the ref on the remote repository, the update hook is invoked. Its exit status determines the success or failure of the ref update.

`post-receive`:
> This hook is invoked by git-receive-pack on the remote repository, which happens when a git push is done on a local repository. It executes on the remote repository once after all the refs have been updated.

There is added safety here by using `post-receive`. `hook` is typically used to determine *whether* something can be pushed, while `post-receive` is for making an action *after* something has been pushed. In this case we only want to update the pull request if the forked repository is updated.

**Comment by Unknwon** [comment]
@JohnMaguire Gogs currently does not delegate (will be https://github.com/gogits/gogs/issues/1623) hooks but directly write content into it.

So if Gogs uses `post-receive`, then users are not able to use this hook themselves anymore, AFAIK many users do use this hook for automations.

**Comment by mhartkorn** [comment]
@Unknwon Did you find time to look into it?

I checked point 3 again and it turns out that Gitlab doesn't protect their `refs/merge-requests/<id>/head` (their equivalent to `refs/pull/<id>/head`) against wrong pushes. So Gogs could do the same in my opinion.

**Comment by Unknwon** [comment]
> I checked point 3 again and it turns out that Gitlab doesn't protect their refs/merge-requests/<id>/head (their equivalent to refs/pull/<id>/head) against wrong pushes. So Gogs could do the same in my opinion.

Sure, it is the easier part in my opinion.

**Comment by mhartkorn** [comment]
Nice, thanks, I'll tick the box above then and leave a comment about it.

**Comment by mhartkorn** [comment]
Sorry if I sound annoying but were you able to review this PR @Unknwon ? I think it's nice to have for Go development because it makes testing PRs easier.

**Comment by Unknwon** [comment]
Will do ASAP. Sorry about that :mask:

**Comment by Unknwon** [comment]
Sorry about the delay, merging now, thanks again!

**Pull request merged**